R/utils.R

```diff
@@ -0,0 +1,146 @@
+# Implements "whitespace eating" functionality on top of a normal TextWriter.
+# (WSTextWriter doesn't inherit from TextWriter, as it presents a different
+# interface.)
+#
+# WSTextWriter relies on the caller distinguishing between writes of important
+# content, and writes of whitespace that may or may not be elided (`.$write()`
+# vs `.$writeWS()`).
+#
+# At any point, `eatWS` may be called, which will cause any recent `writeWS`
+# operations (i.e. those since either the beginning of time, or the most recent
+# `write` operation) to be undone, AND for any future `writeWS` calls to be
+# ignored. A call to `write` will be respected, and will restore normal
+# behavior.
```

Comment:
This API is a little confusing to me. Having to call `eatWS` after every `write` call seems a little tedious and error prone. I haven't looked at the usage enough, so maybe this ends up being simpler for the client, but this is surprising to me.

Would it be cleaner to have the client identify the portions of the text that it's writing out and then let the writer filter/scrub as it sees fit? i.e. 
```
writer$noWSMode(c("before", "inside")
writer$writeWS("\t", tag="before")
writer$resetWSMode()
```

then the writer can just filter out which WS it should be writing and which it shouldn't?